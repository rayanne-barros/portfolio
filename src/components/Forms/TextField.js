import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../foundation/Text';

const InputWrapper = styled.div`
    margin-bottom: 17px;
`;

const Input = styled(Text)`
  width: 100%;
  border: none;
  border-bottom: 1px solid ${({ theme }) => theme.colors.primaryFont};
  padding: 12px 0;
  outline: 0; 
  background-color: ${({ theme }) => theme.colors.secondary};
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'paragraph2',
};

export default function TextField({
  placeholder, name, onChange, value,
}) {
  return (
    <InputWrapper>
      <Input
        type="text"
        name={name}
        placeholder={placeholder}
        onChange={onChange}
        value={value}
        required
      />
    </InputWrapper>
  );
}

TextField.propTypes = {
  placeholder: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
};
