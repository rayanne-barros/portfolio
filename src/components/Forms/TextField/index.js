/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../../foundation/Text';

const InputWrapper = styled.div`
    margin-bottom: 17px;
`;

const Input = styled(Text)`
  width: 100%;
  border: none;
  border-bottom: 1px solid ${({ theme }) => theme.colors.primary};
  padding: 12px 0;
  outline: 0; 
  background-color: ${({ theme }) => theme.colors.secondary};
  border-radius: ${({ theme }) => theme.borderRadius};

  ${({ theme, isFieldInvalid }) => isFieldInvalid && css`
    border-color: ${theme.colors.error};
    & + span {
      color: ${theme.colors.error};
      font-size: 11px;
    }
  `}
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'paragraph2',
};

export default function TextField({
  placeholder, name, onChange, value, type, error, isTouched, ...props
}) {
  const hasError = Boolean(error);
  const isFieldInvalid = hasError && isTouched;
  return (
    <InputWrapper>
      <Input
        type={type}
        name={name}
        placeholder={placeholder}
        onChange={onChange}
        value={value}
        isFieldInvalid={isFieldInvalid}
        required
        {...props}
      />

      {isFieldInvalid && (
        <Text
          variant="smallestException"
          color="error"
          textAlign="left"
          role="alert"
        >
          {error}
        </Text>
      )}
    </InputWrapper>
  );
}

TextField.defaultProps = {
  error: '',
  isTouched: false,
};

TextField.propTypes = {
  placeholder: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  error: PropTypes.string,
  isTouched: PropTypes.bool,
};
