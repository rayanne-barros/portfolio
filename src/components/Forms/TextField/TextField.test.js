import React from 'react';
import user from '@testing-library/user-event';
import { render, screen } from '../../../infra/test/testUtils';
import TextField from './index';

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        placeholder="Nome"
        value="Rayanne"
        onChange={() => { }}
        name="nome"
        type="text"
      />,
    );
    const textField = screen.getByPlaceholderText(/nome/i);
    expect(textField).toMatchSnapshot();
  });

  describe('when field is valid', () => {
    describe('and user is typing', () => {
      test('the value must be updated', () => {
        const onChangeMock = jest.fn();
        render(
          <TextField
            placeholder="Nome"
            value=""
            onChange={onChangeMock}
            name="nome"
            type="text"
            isTouched
          />,
        );
        const inputNome = screen.getByPlaceholderText(/nome/i);
        user.type(inputNome, 'Rayanne');
        expect(onChangeMock).toHaveBeenCalledTimes(7);
      });
    });
  });

  describe('when field is invalid', () => {
    test('displays the respective error message', () => {
      render(
        <TextField
          placeholder="Email"
          value="email.teste.com"
          onChange={() => { }}
          name="email"
          type="email"
          isTouched
          error="O campo email é obrigatório"
        />,
      );
      const inputEmail = screen.getByPlaceholderText(/email/i);
      expect(inputEmail).toHaveValue('email.teste.com');
      expect(screen.getByRole('alert')).toHaveTextContent('O campo email é obrigatório');
      expect(inputEmail).toMatchSnapshot();
    });
  });
});
