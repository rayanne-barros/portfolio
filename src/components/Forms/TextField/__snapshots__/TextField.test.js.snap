// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  font-size: 28px;
  font-weight: 300;
  line-height: 1.25;
  color: #FFFFFF;
}

.c1 {
  width: 100%;
  border: none;
  border-bottom: 1px solid #004AAD;
  padding: 12px 0;
  outline: 0;
  background-color: #000000;
}

<input
  class="c0 c1"
  name="nome"
  placeholder="Nome"
  required=""
  type="text"
  value="Rayanne"
/>
`;

exports[`<TextField /> when field is invalid displays the respective error message 1`] = `
.c0 {
  font-size: 28px;
  font-weight: 300;
  line-height: 1.25;
  color: #FFFFFF;
}

.c1 {
  width: 100%;
  border: none;
  border-bottom: 1px solid #004AAD;
  padding: 12px 0;
  outline: 0;
  background-color: #000000;
  border-color: #dc3545;
}

.c1 + span {
  color: #dc3545;
  font-size: 11px;
}

<input
  class="c0 c1"
  name="email"
  placeholder="Email"
  required=""
  type="email"
  value="email.teste.com"
/>
`;
